blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala

```diff
@@ -48,13 +50,18 @@ object BlazeClient {
               .flatMap { _ => manager.release(connection) }
             Task.now(DisposableResponse(r, dispose))
 
-          case -\/(Command.EOF) =>
+          case -\/(Command.EOF) if !isFresh =>
             invalidate(connection).flatMap { _ =>
-              manager.borrow(key).flatMap { newConn =>
-                loop(newConn, flushPrelude)
+              manager.borrow(key, true).flatMap { newConn =>
+                loop(newConn, flushPrelude, true)
```

Comment:
That is the missed decrement.

Signaling whether the connection is fresh is fine with me but i'd prefer a
case class to a tuple.

I can update this later or someone can make a new one. Unless I hear
otherwise I'll update this one in the evening.
On Mar 17, 2016 12:52 PM, "Ross A. Baker" <notifications@github.com> wrote:

> In blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala
> <https://github.com/http4s/http4s/pull/572#discussion_r56548547>:
>
> >              invalidate(connection).flatMap { _ =>
> > -              manager.borrow(key).flatMap { newConn =>
> > -                loop(newConn, flushPrelude)
> > +              manager.borrow(key, true).flatMap { newConn =>
> > +                loop(newConn, flushPrelude, true)
>
> This is going to be less aggressive in evicting stale connections. We
> might now open a connection when there is a perfectly fine one next in the
> queue. What I have not done is measure how long it takes to loop through
> several stale connections, so this may or may not be a good thing.
>
> With @alissapajer <https://github.com/alissapajer>'s idea on borrow, we
> could loop the old way, and still have the termination condition we need.
> Instead of asking for a fresh one, we'd be told whether it's a fresh one.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/http4s/http4s/pull/572/files/a7fcfa2239212a160b8d061f21ace10d5617d8b8#r56548547>
>